include/envoy/redis/codec.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+
+namespace Redis {
+
+/**
+ * All RESP types as defined here: https://redis.io/topics/protocol
+ */
+enum class RespType { Null, SimpleString, BulkString, Integer, Error, Array };
+
+/**
+ * A variant implementation of a RESP value optimized for peformance. A C++11 union is used for
+ * the underlying type so that no unnecessary allocations/constructions are needed.
+ */
+class RespValue {
```

Comment:
In 99% of cases, yes. I'm will to compromise in certain cases mostly as it relates to performance. (We did something similar with HTTP headers). RespValue is guaranteed to become a performance bottleneck for the redis code. In order to use unions and allow vectors that contain actual RespValue vs. pointers, we need a concrete class and not an interface. So in this particular case I think it's worth it to have an exception to the general rule.